### DISCUSSION TITLE
Custom Screens

### DISCUSSION THREAD

**Discussion by peepshow-21** [blakadder/nspanel#16] [General]
Is is possible to send the custom TFT files to the screen via this Tasmota implementation?

If so, I guess that is all that's needed and a custom ".be" files to map that to the MTQ messages, like the current one.

Importantly, can the original TFT file be put back, when tinkering is over, without having to hardware reflash the whole firmware?

The place to be would be keeping Tasmota on, and just be able to drop differnt TFT's to the screen.

**Comment by blakadder** [comment]
It has not been attempted or tested.

You can flash back the original display firmware same as you can flash custom firmware

**Reply by joBr99** [reply]
This would also be needed in case there are new versions of the stock tft firmware. (And a tasmota user want's to upgrade to the newest version of this)

**Reply by peepshow-21** [reply]
I've only used Tasmota and it's what I have installed on all my devices, but I've seen in other threads people getting custom screens using ESPHome. Will I need switch from Tamota to ESPHome if I want to look at custom screens then? :/

I don't really understand to the detail of you, so sorry if I have to ask questions to understand.

If I flash tasmota.bin as per the guides, it doesn't change the firmware in screen, just the ESP side of it, right? In the same folder where tasmota.bin is there is a nspanl origninal firmware bin too. I assumed this was just the same ESP side as the tasmota side (no screen firmware). Are you saying there is a screen-firmware.bin part too, which I flash exactly the same way as the tasmoto.bin? From the little I read, there seems to be an original.tft floating about, but I assumed that wasn't put on like firmware (tasmota.bin), but must be some other way. Could you clear this up, as with my level on knowledge (still learning as I go), the changes of me bricking my box are not zero!! :)

**Reply by peepshow-21** [reply]
Looks like this is the thing required here;
https://nextion.tech/2017/12/08/nextion-hmi-upload-protocol-v1-1/
But that's way outside my scrope! lol

Hope someone smarter than me can do it. But looks doable. I think it would mean copying a file to the ESP side with the File Manager console the issing a tasmota command to send that file using the above defined protocol to the screen via it's serial link.

One question though, would this need to be done in tasmoa core code, or could this be done via the berryscript (.be) files? I've never seem this before but it looks like a programming language to some degree which has access to the serial port. Is it at a level where it could handle this kind of job? 

Does anyone know if anyone is looking at this? maybe the people that did the orignal nspaben.be ?

**Reply by joBr99** [reply]
@blakadder did the original nspanel.be :)

The nspanel has two "parts" the esp32 which can be flashed with tasmota, which controls the screen. With tasmota you cannot flash a new tft file onto the screen, so you have to stick with the stock screen firmware for now.

But you are right, HMI Upload Protocol needs to be implemented to flash tft files onto the screen.

https://github.com/jesserockz/esphome/blob/nspanel/component/esphome/components/nextion/nextion_upload.cpp
https://unofficialnextion.com/t/nextion-upload-protocol-v1-2-the-fast-one/1044/2

I guess it would be possible to implement this in tasmota core or in berry, but as far as I can see is the current berry implementation lacking some feature's that you would need to implement the HMI Protocol. (But I'm not a tasmota dev and not very familar with berry.)

**Reply by peepshow-21** [reply]
I had a little look at berry and it seems it might, there are already the serial comms to the screen, and it  looks like the language has an open() command for files, which return a file, which has a read(byte) on it. so I'm guess that could be used to do this. I've never seen berry until an hour ago, so just guessing for now. but wouldn't seem to hard it it all works as it reads it should. I might have a go over the weekend. 

looks like berry has a webclient class too, so could maybe even do;
```
cmnd/nspanel/flash http://192.168.0.0/screen.tft
```
and have berry in the driver.to fulfill that

You's need a different nspanel.be after though, that could comunicate with the new screen firmware

**Reply by joBr99** [reply]
Take a look into the nextion upload implementation of esphome, the tft files are (or can be) bigger than the space that you have available on the flash/memory of the esp32. So you need to request parts of the file via http (Range Header) and send thoose to the display.

But there are multiple limitations that need to be solved before you can use the webclient in berry for a implemenation of the upload protocol. Copy-pasted them from tasmota doc, my comments in []
- No access to response headers [Needed to get the size of the tft file, must be send to the display at the start of flashing]
- Only supports text responses (html, json...) but not binary content yet (no NULL char allowed) [tft file is binary]

I also played a bit with the tcp class in tasmota berry, but it crashed my esp after calling tcp.available() or 
tcp.connected().

**Reply by peepshow-21** [reply]
It's never as much fun, if it's too easy! :)

Yes, i saw it needs sent to the screen in 4096 chunks. Maybe encoding the file before sending, so not binary would work. Read the file via webclient twice, once to size, once to send.

Some think you can work around, it's more stuff like it not being able to send 4096 (if it were the case) or things with no workaround the ruin things! :)

Probably worth a try though.

**Reply by peepshow-21** [reply]
@joBr99 , looking at berry and I don't understand this in webclient;

read | (addr:int, reg:int, size:int) -> int or nilRead a value of 1..4 bytes from address addr and register reg. Returns nil if no response.

what is addr and reg? no idea how to use this fuunction. help?

**Reply by peepshow-21** [reply]
I've almost got it working! :)

Should have an example tomorrow :)

**Comment by peepshow-21** [comment]
Success!!!

I now have working custom screen flashing for tasmota! :)

![image](https://user-images.githubusercontent.com/42150988/149673844-8ca5700d-2780-4151-a5a0-1409fde65d15.png)

You simply put the TFT on a accessable URL and run this command.

There is a small java program to run on the host to convert the TFT into a format the can be loaded as needed to work with some of the issues of using Berry. But you just need the standard tasmota binary and a new `nextion.bc` script.

I've tried flashing a test custom TFT and then flashing the stock one back. Worked fine.

Customer screens here we come.

I'll tidy it up a bit then post for those interested

Thanks to @blakadder for the ordinal script. I've not heard of berry until 2 days ago so would never have got a solution as fast without that as a starting point. Wish he'd tell me how to get my buzzer going though! :)

**Reply by joBr99** [reply]
That's really nice, guess I will ditch my esphome component than and will go back to tasmota for my custom ui.
~~Did you already publish your berry script and the java application somewhere?~~
Take your time tidy up your code, didn't read that part :)

Did you implement Upload Protocol 1.1 or 1.2?

**Comment by peepshow-21** [comment]
Early version here;

https://github.com/peepshow-21/ns-flash

Let me know how it goes!

I just did 1.1 for now as it was the first doc I saw. Will look at that next, I understand it's faster.
It's not the best way I did it, but was limited dur to the things that worked in berry!

**Reply by joBr99** [reply]
Can you also push the code of your ns-flash.jar? Would like to compile myself and not execute java bytecode of a random person on the internet :)

Btw, are you on the unoffical nextion discord?

**Reply by peepshow-21** [reply]
Source;
https://github.com/peepshow-21/ns-flash 

no, I'm not on discord at all

**Reply by puca11** [reply]
@peepshow-21 have you tried setoption 111 for the buzzer, was clicking for me but setting it on fixed it (well at least it beeps).

**Reply by peepshow-21** [reply]
worked a treat thanks! :)

**Comment by peepshow-21** [comment]
**NxPanel - Full featured replacement custom firmware for NsPanel**

Features

* Custom button screens
* All buttons labelled (not 'outlet1...)
* All buttons have icon choice
* Unlimited depth to navigation
* Dim/Sleep configuration
* Swipe support
* Notifications
* Home screen warnings
* Alarm panel
* Thermostat
* Status Panel
* Alerts
* Standard dimmer
* Color & CT dimmer
* Media Player Panel
* Nextion 2.1 fast flashing
* Auto updates - no manual flash after initial install
* Backend automation agnostic
* Communications - JSON over MQTT

[Support Thread](https://community.openhab.org/t/nxpanel-replacement-firmware-for-sonoff-nspanel/132869)

**Comment by kptkip** [comment]
@blakadder Is it possible to change the Icons from Widgets with own PNG/SVG/JPG-Files?

**Reply by joBr99** [reply]
No, you cannot modify the stock display firmware, you have to build your own or use one of the existing projects.

**Comment by eokgnah** [comment]
Has someone please an example custom nextion.be?
I understand how to make a custom TFT and how to upload it.
I understand the original nextion.be.

But i don't see how or where the (tasmota) values like "temperature" are connected/translated to the (nextion) Page/ID to be shown.

So my experiment to make a own tft to show the values "temperature/tmin/tmax/..." like the original one was not showing any values with the stock tasmota.be

Or did i misunderstand something?

...Eok

**Reply by joBr99** [reply]
You need to implement a serial parser in nextion, take a look at my berry script and the HMI Project

https://github.com/joBr99/nspanel-lovelace-ui

**Comment by eokgnah** [comment]
Thanks, i looked at lovelace and tried it. but without luck.
Maybe i ask the wrong things.

Ok, for example.

1. i installed tasmota (OK)
2. i made my own HMI/TFT with NextionEditor (OK)
3. i installed the tft on the display (OK)

On (for example) Page "0" i have a text field with ID "3" and objname "temperature"
Where and how do i send values to this 0/3 from the console or later over mqtt?

Without homeassistant or anything else.
Just my HMI and tasmota and mqtt to fill the pages with data.

I see some parsing in the .be, but i don't get where the "switch" from the variable "temperature" to "0/3" happens.

I thought i just have to know the page/id which i want to fill and then send something there to see it on the screen?
So i thought there might somewhere a table with "name" <-> "page/id". But i don't find it.
Or is my understanding about this part wrong?

If i understand it with one value, i can do it with any. It feels like just a small step to do. 
But i don't see what you are trying to show me at the moment. :-(

**Reply by joBr99** [reply]
In the stock nspanel firmware there is a Reparse Protocol implemented you are sending a "json" to the nextion display, the display is parsing it and then it is assigning the stuff to the text field. (It's not really json, because nextion language is not capable to parse json, but thats a long story)

To modify the text field you need to speak the nextion instruction language with the panel, so you need to send something like "textField0=test123", but this is way to complex to handle with berry and tasmota. 

Just implement Reparse Protocol in Nextion, this allows you to define your own commands and simplifies things a lot. 
See this to get an idea for the commands I used: https://github.com/joBr99/nspanel-lovelace-ui/tree/main/HMI. (Do not try to implement json like sonoff did, this is just dumb.)
The parser in nextion is implemented on each page in the Timer tmSerial.
Futhermore, my project isn't necessary limited to HomeAssistant, you just need a backend that is controlling the panel with the commands.

**Reply by eokgnah** [reply]
Ok, i found the scripts in tmSerial.
Since i don't use a backend im still somewhat clueless how all comes together.
It's quite complex with all the timers, pages and thoughts you programmed into it.
Way more then i need (or understand) at the moment.

Can you hint me a mqtt message which i just send and see the value on the display, then i can see something happening and can piece the way back and alter to fit my need.

Im still stuck with the spinner page with no clue how to format a mqtt message.

A plain simple HMI with just 1 page and maybe 1 number, one text, one button on it with associated mqtt messages as example would be top for all who want to do something similar.

All the pages and interacting code is cool, but at the moment a bit complex to understand.
At least at a start.

Can i ask for this simple 1 page / mqtt message example project for all of us? ;-)
Looked around, but did not find such a starter for newbies.
Would be cool.

**Reply by joBr99** [reply]
If you are in the waiting for content screen it will periodically send this message: `event,startup,39,eu`

To get out of this screen you have to navigate away from this page, Try the following commands in tasmota console. (You can also send it over mqtt to the CutomSend Topic)

**time/date are stored in a global variable, but only used on screensaver**
`CustomSend time~09:07~`
`CustomSend date~Samstag, 6. August 2022`
**timeout is a timer on all pages (execpt the screensaver that will send a message after X seconds to allow the backend to naviagte back to the screensaver)**
`CustomSend timeout~20`
**settings for brightness (and background color)**
`CustomSend dimmode~10~100~0`
**change the current page to the screensaver**
`CustomSend pageType~screensaver`
**update the weather on screensaver**
`CustomSend weatherUpdate~X~1C~So.~Y~2C~Mo.~X~2C~Di.~Z~3C~Mi.~A~4C~~`

If you are touching the screensaver the firmware will send the following message: `event,buttonPress2,screensaver,bExit,1`

**change current page to cardEntities**
`CustomSend pageType~cardEntities`
**set some content of cardEntities (stipped this a bit down, also I replaced the icons in this command, so it works in tasmota console)**
`CustomSend entityUpd~Title~1|1~shutter~cover.rolladenfenster_cover_1~A~17299~E1 Name~X|Y|Z|disable|enable|enable~`

Which HomeAutomation System are you using? If you are planing to implement a backend for a new HA System, I'm going to  help you understand the serial protocol of this firmware, it can be confusing from time to time, because a lot of stuff is packend into the entityUpd commands to save computing time on nextion display.

**Reply by eokgnah** [reply]
I just use MQTT in the center and for some logic NodeRed and some shell-scripts.

So i hoped for a kind of "standalone" solution. 
After all, thats what tasmota is famous for. Simple, and with no need for a backend.

So i thought sonoff is the right choice.
Just a simple custom HMI which works on the display.
And some mqtt-messages to fill the values or set the icons.
In my simplified mind, i aimed for something like:
_mosquitto_pub -t "NSDisplay/cmnd" -m "temperature:23.7"_
To set the temperature value on the display.
Nothing more, nothing less.
I dont need/want a new backend.
Just some simple mqtt messages which display things.

But maybe thats not how it works?
It feels like i am so close, but still so far.

Mabe i have to look for something else than tasmota to talk to the hmi over mqtt.
But what?

**Reply by eokgnah** [reply]
I'm in...
I see the time and date.

sending:
CustomSend date~Samstag, 6. August 2022
on the console shows it on the screen!

Exactly what i want.
So now i have to search for it and delete all the other code to don't be confused and get it down to the bare minimum to reach this task.
So it is more like:
_mosquitto_pub -t "NSDisplay/cmnd/CustomSend" -m "date~Samstag, 6. August 2022"_
Which is totally perfect and fine!
And if that works, buttons.
I have a Picture of my flat. Rendered in a 3d-programm. With Doors and Windows in red/green according to their status.
I want to have this as a Background on the Display, showing the status.
Looks like im getting there after all. 
I'm again a bit confident to reach this goal at the moment. ;-)

**Reply by joBr99** [reply]
The berry script is just translating mqtt messages to serial. You are setting the content over mqtt with thoose ~ sperated commands. But you have to set the whole content of the page with one command (usally entityUpd in my firmware), because otherwise you will have really long loading times.
```
mosquitto_pub -t "NSDisplay/cmnd/CustomSend" -m "dimmode~20~100~0"
mosquitto_pub -t "NSDisplay/cmnd/CustomSend" -m "pageType~screensaver"
mosquitto_pub -t "NSDisplay/cmnd/CustomSend" -m "time~XX:XX"
```
I started with a nodered backend (with the HA node module), but this was too complicated, so I switched to AppDaemon.

Last Firmware compatible with nodered version: https://nspanel.pky.eu/lui-nodered.tft
NodeFlow: https://github.com/joBr99/nspanel-lovelace-ui/tree/v1.5.0/nodered

-------

If you want to build such a page, start with modifying one of the existing card like cardQR, just change the gui elements and then modify the entityUpd command to your needs.

From the waiting for content screen you can get there with `pageType~cardQR` If you are a litte bit more expienced you can also add your own page to the pageType command on the other pages (eg the startup/waiting for content page)

**Reply by eokgnah** [reply]
https://github.com/eokgnah/NSPanel-Nextion/
found a cheat sheet with all possible instructions and made some examples.
thanks for the help to get me there.
